test/SILOptimizer/mem2reg_ossa.sil

```diff
@@ -0,0 +1,978 @@
+  %7 = apply %6(%5) : $@convention(thin) (@guaranteed Builtin.BridgeObject) -> ()
+  dealloc_stack %1 : $*AnyObject                  // id: %8
+  destroy_value %5 : $Builtin.BridgeObject               // id: %9
+  %10 = tuple ()                                  // user: %11
+  return %10 : $()                                // id: %11
+} // end sil function 'load_take_use'
+
+sil hidden [ossa] @test : $@convention(method) (@guaranteed Builtin.BridgeObject) -> () {
+bb0(%1 : @guaranteed $String, %2: @owned $Builtin.BridgeObject):
+  %8 = alloc_stack $String
+  %9 = copy_value %1 : $String
+  store %9 to [init] %8 : $*String
+  
+  %18 = alloc_stack $Builtin.BridgeObject
+  store %2 to [init] %18 : $*Builtin.BridgeObject
+  
+  br bb1
+
+bb1:
+  cond_br undef, bb2, bb3
+
+bb2:
+  %35 = load [copy] %8 : $*String
+  destroy_value %35 : $String
+  %42 = copy_value %1 : $String
+  store %42 to [assign] %8 : $*String
+  br bb1
+
+bb3:
+  destroy_addr %18 : $*Builtin.BridgeObject
+  destroy_addr %8 : $*String
+  dealloc_stack %18 : $*Builtin.BridgeObject
+  dealloc_stack %8 : $*String
+  %55 = tuple ()
+  return %55 : $()
+}
+
+// Check that we remove load copy *and* destroy_value. Also after dealloc.
+sil hidden [ossa] @test : $@convention(method) (@guaranteed __CocoaDictionary) -> () {
+bb0(%1 : @guaranteed $__CocoaDictionary):
+  %4 = struct_extract %1 : $__CocoaDictionary, #__CocoaDictionary.object // user: %66
+  %65 = alloc_stack $AnyObject                    // users: %67, %71, %70, %68
+  %66 = copy_value %4 : $AnyObject                // user: %67
+  store %66 to [init] %65 : $*AnyObject           // id: %67
+  %68 = unchecked_addr_cast %65 : $*AnyObject to $*_NSDictionary // user: %69
+  %69 = load [copy] %68 : $*_NSDictionary         // users: %73, %74, %79
+  destroy_addr %65 : $*AnyObject                  // id: %70
+  dealloc_stack %65 : $*AnyObject                 // id: %71
+  %74 = begin_borrow %69 : $_NSDictionary
+  end_borrow %74 : $_NSDictionary
+  destroy_value %69 : $_NSDictionary
+  %55 = tuple ()
+  return %55 : $()
+}
+
+// Same as above but two blocks.
+sil hidden [ossa] @test : $@convention(method) (@guaranteed __CocoaDictionary) -> () {
+bb0(%1 : @guaranteed $__CocoaDictionary):
+  %4 = struct_extract %1 : $__CocoaDictionary, #__CocoaDictionary.object // user: %66
+  %65 = alloc_stack $AnyObject                    // users: %67, %71, %70, %68
+  %66 = copy_value %4 : $AnyObject                // user: %67
+  store %66 to [init] %65 : $*AnyObject           // id: %67
+  %68 = unchecked_addr_cast %65 : $*AnyObject to $*_NSDictionary // user: %69
+  %69 = load [copy] %68 : $*_NSDictionary         // users: %73, %74, %79
+  destroy_addr %65 : $*AnyObject                  // id: %70
+  dealloc_stack %65 : $*AnyObject                 // id: %71
+  %74 = begin_borrow %69 : $_NSDictionary
+  end_borrow %74 : $_NSDictionary
+  br bb1
+  
+bb1:
+  destroy_value %69 : $_NSDictionary
+  %55 = tuple ()
+  return %55 : $()
+}
+
+// TODO:
```

Comment:
I've got these all in the form of hundred line stdlib reproducers. Today (hopefully) I'll turn them into nice tests (like the ones above).